Plugins overview and basic content plugin example

robbiejackson opened the pull request:
I'd appreciate a careful review of this content as I've just been finding out about some of these aspects while writing the documentation.

HLeithner commented on versioned_docs/version-4.4/building-extensions/plugins/basic-content-plugin.md:
````diff
@@ -0,0 +1,291 @@
+# Manifest File
+For general information on manifest files see [Manifest Files](https://docs.joomla.org/Manifest_files).
+
+**plg_shortcodes/shortcodes.xml**
+```xml
````
```suggestion
```xml title="plg_shortcodes/shortcodes.xml"
```

HLeithner reviewed (commented):
I would suggest to use an ascii art for the directory tree, that's easier to adapt in the future

https://asciiflow.com/ can help you to draw it.

For the other diagram maybe a mermaid supported diagram would make it easier for the future.
https://mermaid.js.org/

I didn't read thru the document if it's correct yet but please add blank lines between elements like headline and content or codeblocks, that would make it much easier to read.

Fedik commented on versioned_docs/version-4.4/building-extensions/plugins/basic-content-plugin.md:
```diff
@@ -0,0 +1,291 @@
+    public static function getSubscribedEvents(): array
+    {
+        // we want this plugin to work only on the site, not eg on administrator or api
+        if (Factory::getApplication()->isClient('site')) {
```
Please do not use it here, or better remove this example, it is a bad one :smile: 
I know Joomla have it in the core, but it not a good use.

In general the method `getSubscribedEvents` should not depend from current runtime.

Fedik replied:
Alternatively add it in to the listeners like:
```php
if ($this->getApplication()->isClient('site')) {
  return;
}
```

robbiejackson replied:
@Fedik @HLeithner Thanks for taking the time to review! I'm just rather surprised at the suggestion to move the client check from getSubscribedEvent. Surely it's more performant to do it there, so that Joomla doesn't waste time triggering events that are going to be ignored anyway, isn't it? 

If there's a good reason to do it in the listener then it would be worth adding an explanation in the document as to why it should be done there instead.

Fedik replied:
What I mean, we should not to advertise a bad practice in our documentation, even if Joomla use it in own code :wink: 

It do not take much time to trigger the event, I would not wory about it. 
This not realy about perfomance but more about code reliability.

When in future we find a better soluton, then we can update the manual. But for now better not to advertise such thing.

Fedik replied:
The root in how the events works, the listener initialisation:
```php
$subscriber = new EventSubscriber(); // Our plugin
$dispatcher->addSubscriber($subscriber); // Call $subscriber->getSubscribedEvents() internally
```
This code will initialise new Subscriber quickly.

Keeping the plugin `__constructor` and `getSubscribedEvents` light will keep whole booting process fast, and code more reliable.

In fact, not all listeners from subscriber is called in the end. 
Everything depend from plugin group and current runtime, example `onContentPrepare` will never be called if current component do not need it. Or `onBeforeCompileHead` will never be called if request not `html` and so on.

robbiejackson replied:
ok, I've changed that. Agreed it would be slightly faster for the site (which is the most important), even though it would be a bit slower for the administrator in having to add the listeners.

Fedik replied:
Thanks!

robbiejackson commented:
Overall I've made a number of changes since the initial version:

1. Curly brackets are escaped
2. Blank lines around code blocks
3. title="..." associated with code blocks of files
4. Improved plugin overview diagram, plus source for the diagram
5. Moved client check to listeners in the plugin code
6. Fixed broken link messages

Regarding the diagrams:

- The image showing the file structure isn't a drawing which I've generated, it's just a screenshot. Easier to do that.
- I looked at mermaid but it wasn't sophisticated enough to generate the sort of diagram I wanted. However, the aspect of maintaining diagrams is important, so I've uploaded the source odg file as well (I use openoffice draw).

Fedik commented:
Here is a sugestions for custom graph visualisations.

**Plugins structure**
``` 
flowchart TD
    /plugins-->/system
    /plugins-->/content
    /plugins-->/typeX["Another type"]
    /system-->plugin1
    /system-->plugin2
    /system-->pluginX["..."]
    /content-->plugin4
    /content-->plugin5
    /content-->pluginY["..."]
    /typeX-->plugin6
    /typeX-->plugin7
    /typeX-->plugin8["..."]
```
```mermaid
flowchart TD
    /plugins-->/system
    /plugins-->/content
    /plugins-->/typeX["Another type"]
    /system-->plugin1
    /system-->plugin2
    /system-->pluginX["..."]
    /content-->plugin4
    /content-->plugin5
    /content-->pluginY["..."]
    /typeX-->plugin6
    /typeX-->plugin7
    /typeX-->plugin8["..."]
```

**Process of loading and dispatching**
```
sequenceDiagram
    Code->Dispatcher: Get Dispatcher <br>$dispatcher
    Code->>PluginHelper: PluginHelper::importPlugin($type, null, true, $dispatcher)
    PluginHelper-->>Plugins Table: Query all active
    Note over PluginHelper, Plugins Table: Only once at runtime, and <br>result is cached in memory
    Plugins Table-->>PluginHelper: Return all active
    PluginHelper-->>Dispatcher: Import only <br>group of $type <br>(once at runtime)
    Dispatcher-->>Plugin: Call $plugin->getSubscribedEvents() <br>for each plugin in group
    Plugin-->>Dispatcher: Return list of listeners
    Code->>Dispatcher: dispatch('onFooBar', <br>$eventObject)
    Dispatcher->>Plugin: Call listener for onFooBar event
    Note over Plugin: Plugin executes own logic <br>and, when event support this, <br>add result to $eventObject
```
```mermaid
sequenceDiagram
    Code->Dispatcher: Get Dispatcher <br>$dispatcher
    Code->>PluginHelper: PluginHelper::importPlugin($type, null, true, $dispatcher)
    PluginHelper-->>Plugins Table: Query all active
    Note over PluginHelper, Plugins Table: Only once at runtime, and <br>result is cached in memory
    Plugins Table-->>PluginHelper: Return all active
    PluginHelper-->>Dispatcher: Import only <br>group of $type <br>(once at runtime)
    Dispatcher-->>Plugin: Call $plugin->getSubscribedEvents() <br>for each plugin in group
    Plugin-->>Dispatcher: Return list of listeners
    Code->>Dispatcher: dispatch('onFooBar', <br>$eventObject)
    Dispatcher->>Plugin: Call listener for onFooBar event
    Note over Plugin: Plugin executes own logic <br>and, when event support this, <br>add result to $eventObject
```

robbiejackson commented:
Thanks for your diagram suggestions @Fedik but I'm afraid I still prefer my own. The reason is that I wrote this page for someone who was completely new to Joomla plugins, and didn't know what they were. So I really want a diagram to convey in simplest terms what plugins are and how they work. The sequence diagram is rather too much detail for that.

Fedik commented:
Yeah,  need to know how to read this sequenceDiagram (follow arrows, left to right, top to bottom). 
You can still put it somewhere at bottom, there still many expirianced Joomla users who have no idea how joomla plugins works.
In general I tried to follow your graph. 

One more note, one misleading thing on your image is `importPlugin() -> get subfolders`. This can make User to think that Joomla do plugin search  in file system directly. But what actualy importPlugin() doing is loading them from Database, and then looking for plugin files.

robbiejackson commented:
Thanks @Fedik all good suggestions, which I'll implement.

I just can't see the preview of the mermaid diagram. I presume I have to install an extension onto github dev, but can you tell me which extension? There seem to be a few mermaid ones.

Fedik commented:
For "github dev" (and VCode) there a plugin "Markdown Preview Mermaid Support"

HLeithner commented:
I use https://mermaid.live

HLeithner commented:
I'm merging this for now, we can always improve it.
